File: core/baker-wirespec/src/main/java/com/ing/baker/recipe/wirespec/BakerJavaEmitter.java
package com.ing.baker.recipe.wirespec;

import community.flock.wirespec.compiler.core.emit.LanguageEmitter;
import community.flock.wirespec.compiler.core.emit.FileExtension;
import community.flock.wirespec.compiler.core.emit.Shared;
import community.flock.wirespec.compiler.core.parse.ast.Channel;
import community.flock.wirespec.compiler.core.parse.ast.Endpoint;
import community.flock.wirespec.compiler.core.parse.ast.Enum;
import community.flock.wirespec.compiler.core.parse.ast.Field;
import community.flock.wirespec.compiler.core.parse.ast.Identifier;
import community.flock.wirespec.compiler.core.parse.ast.Module;
import community.flock.wirespec.compiler.core.parse.ast.Reference;
import community.flock.wirespec.compiler.core.parse.ast.Refined;
import community.flock.wirespec.compiler.core.parse.ast.Type;
import community.flock.wirespec.compiler.core.parse.ast.Union;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class BakerJavaEmitter extends LanguageEmitter {

    @NotNull
    @Override
    public String getSingleLineComment() {
        return "//";
    }

    @NotNull
    @Override
    public FileExtension getExtension() {
        return FileExtension.Java;
    }

    @Nullable
    @Override
    public Shared getShared() {
        return null;
    }

    @NotNull
    @Override
    public String notYetImplemented() {
        return "";
    }

    @NotNull
    @Override
    public String emit(@NotNull Identifier identifier) {
        return identifier.getValue();
    }

    @NotNull
    @Override
    public String emit(@NotNull Endpoint endpoint) {
        String name = emit(endpoint.getIdentifier());
        String interactionName = name + "Interaction";
        String outcomeName = name + "Outcome";

        StringBuilder sb = new StringBuilder();

        // Imports
        sb.append("import com.ing.baker.recipe.javadsl.Interaction;\n");
        sb.append("import com.ing.baker.recipe.annotations.FiresEvent;\n\n");

        // Interface
        sb.append("public interface ").append(interactionName).append(" extends Interaction {\n");
        sb.append("    interface ").append(outcomeName).append(" {}\n");

        // Response events — one per status code
        for (Endpoint.Response response : endpoint.getResponses()) {
            String eventName = name + "Response" + response.getStatus();
            if (response.getContent() != null) {
                String bodyType = emitReference(response.getContent().getReference());
                sb.append("    class ").append(eventName)
                  .append(" implements ").append(outcomeName).append(" {\n");
                sb.append("        public final ").append(bodyType).append(" body;\n");
                sb.append("        public ").append(eventName).append("(").append(bodyType).append(" body) { this.body = body; }\n");
                sb.append("    }\n");
            } else {
                sb.append("    class ").append(eventName)
                  .append(" implements ").append(outcomeName).append(" {}\n");
            }
        }

        sb.append("\n");

        // @FiresEvent annotation
        String responseClasses = endpoint.getResponses().stream()
            .map(r -> name + "Response" + r.getStatus() + ".class")
            .collect(Collectors.joining(", "));
        sb.append("    @FiresEvent(oneOf = {").append(responseClasses).append("})\n");

        // apply() method — collect all input params
        List<String[]> params = collectParams(endpoint);
        String paramList = params.stream()
            .map(p -> p[1] + " " + p[0])
            .collect(Collectors.joining(", "));
        sb.append("    ").append(outcomeName).append(" apply(").append(paramList).append(");\n");
        sb.append("}\n");

        return sb.toString();
    }

    @NotNull
    @Override
    public String emit(@NotNull Type type, @NotNull Module module) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emit(@NotNull Type.Shape shape) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emit(@NotNull Field field) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emit(@NotNull Reference reference) {
        return emitReference(reference);
    }

    @NotNull
    @Override
    public String emit(@NotNull Reference.Primitive.Type.Constraint constraint) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emit(@NotNull Enum anEnum, @NotNull Module module) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emit(@NotNull Union union) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emit(@NotNull Refined refined) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emitValidator(@NotNull Refined refined) {
        return notYetImplemented();
    }

    @NotNull
    @Override
    public String emit(@NotNull Channel channel) {
        return notYetImplemented();
    }

    /**
     * Maps wirespec Reference types to Java type names.
     */
    private String emitReference(Reference reference) {
        String typeName;
        if (reference instanceof Reference.Primitive primitive) {
            typeName = switch (primitive.getType()) {
                case Reference.Primitive.Type.String s -> "String";
                case Reference.Primitive.Type.Integer i -> switch (i.getPrecision()) {
                    case P32 -> "Integer";
                    case P64 -> "Long";
                };
                case Reference.Primitive.Type.Number n -> switch (n.getPrecision()) {
                    case P32 -> "Float";
                    case P64 -> "Double";
                };
                case Reference.Primitive.Type.Boolean b -> "Boolean";
                case Reference.Primitive.Type.Bytes b -> "byte[]";
                default -> "Object";
            };
        } else if (reference instanceof Reference.Custom custom) {
            typeName = custom.getValue();
        } else if (reference instanceof Reference.Iterable iterable) {
            typeName = "java.util.List<" + emitReference(iterable.getReference()) + ">";
        } else if (reference instanceof Reference.Dict dict) {
            typeName = "java.util.Map<String, " + emitReference(dict.getReference()) + ">";
        } else if (reference instanceof Reference.Unit u) {
            typeName = "Void";
        } else {
            typeName = "Object";
        }
        return typeName;
    }

    /**
     * Collects all input parameters from path segments, query params, headers, and request body.
     * Returns list of [name, type] pairs.
     */
    private List<String[]> collectParams(Endpoint endpoint) {
        List<String[]> params = new ArrayList<>();

        // Path parameters
        for (Endpoint.Segment segment : endpoint.getPath()) {
            if (segment instanceof Endpoint.Segment.Param param) {
                params.add(new String[]{
                    param.getIdentifier().getValue(),
                    emitReference(param.getReference())
                });
            }
        }

        // Query parameters
        for (Field query : endpoint.getQueries()) {
            params.add(new String[]{
                query.getIdentifier().getValue(),
                emitReference(query.getReference())
            });
        }

        // Header parameters
        for (Field header : endpoint.getHeaders()) {
            params.add(new String[]{
                header.getIdentifier().getValue(),
                emitReference(header.getReference())
            });
        }

        // Request body
        for (Endpoint.Request request : endpoint.getRequests()) {
            if (request.getContent() != null) {
                params.add(new String[]{
                    "body",
                    emitReference(request.getContent().getReference())
                });
            }
        }

        return params;
    }
}

File: core/baker-wirespec/src/main/kotlin/com/ing/baker/recipe/wirespec/BakerJavaEmitter.kt
package com.ing.baker.recipe.wirespec

import arrow.core.NonEmptyList
import arrow.core.toNonEmptyListOrNull
import community.flock.wirespec.compiler.core.emit.Emitted
import community.flock.wirespec.compiler.core.emit.FileExtension
import community.flock.wirespec.compiler.core.emit.PackageName
import community.flock.wirespec.compiler.core.emit.Shared
import community.flock.wirespec.compiler.core.parse.ast.AST
import community.flock.wirespec.compiler.core.parse.ast.Channel
import community.flock.wirespec.compiler.core.parse.ast.Endpoint
import community.flock.wirespec.compiler.core.parse.ast.Enum
import community.flock.wirespec.compiler.core.parse.ast.Module
import community.flock.wirespec.compiler.core.parse.ast.Reference
import community.flock.wirespec.compiler.core.parse.ast.Refined
import community.flock.wirespec.compiler.core.parse.ast.Type as WsType
import community.flock.wirespec.compiler.core.parse.ast.Union
import community.flock.wirespec.compiler.utils.Logger
import community.flock.wirespec.ir.core.File
import community.flock.wirespec.ir.core.Type
import community.flock.wirespec.ir.core.file
import community.flock.wirespec.ir.core.`interface`
import community.flock.wirespec.ir.emit.IrEmitter
import community.flock.wirespec.ir.generator.Generator
import community.flock.wirespec.ir.generator.JavaGenerator

class BakerJavaEmitter(
    private val packageName: PackageName,
) : IrEmitter {

    constructor() : this(PackageName(""))

    override val extension: FileExtension = FileExtension.Java
    override val shared: Shared? = null
    override val generator: Generator = JavaGenerator

    private var currentModule: Module? = null

    internal fun setModule(module: Module) {
        currentModule = module
    }

    override fun emit(ast: AST, logger: Logger): NonEmptyList<Emitted> {
        val files = ast.modules.toList().flatMap { module ->
            currentModule = module
            module.statements.filterIsInstance<Endpoint>().map { emit(it) }
        }
        val nel = files.toNonEmptyListOrNull()
            ?: return NonEmptyList(Emitted("empty." + extension.value, ""), emptyList())
        return nel.map { f -> Emitted(buildFileName(f), generator.generate(f)) }
    }

    private fun buildFileName(file: File): String {
        val dir = packageName.takeIf { it.value.isNotEmpty() }
            ?.let { it.value.replace('.', '/') + "/interaction/" }
            ?: ""
        return dir + file.name.value() + "." + extension.value
    }

    override fun emit(endpoint: Endpoint): File {
        val name = endpoint.identifier.value
        val interactionName = "${name}Interaction"
        val outcomeName = "${name}Outcome"
        val params = collectApplyParams(endpoint)
        val responseClassNames = endpoint.responses.map { "${name}Response${it.status}" }

        return file(interactionName) {
            packageName.takeIf { it.value.isNotEmpty() }
                ?.let { `package`("${it.value}.interaction") }

            import("com.ing.baker.recipe.javadsl", "Interaction")
            import("com.ing.baker.recipe.annotations", "FiresEvent")

            `interface`(interactionName) {
                extends(type("Interaction"))
                `interface`(outcomeName)

                endpoint.responses.forEach { response ->
                    val eventName = "${name}Response${response.status}"
                    val content = response.content
                    if (content != null) {
                        struct(eventName) {
                            implements(type(outcomeName))
                            field("body", content.reference.toIrType())
                        }
                    } else {
                        struct(eventName) {
                            implements(type(outcomeName))
                        }
                    }
                }

                val firesEventArgs = responseClassNames.joinToString(", ") { "$it.class" }
                raw("@FiresEvent(oneOf = {$firesEventArgs})")

                function("apply") {
                    returnType(type(outcomeName))
                    params.forEach { (n, t) -> arg(n, t) }
                }
            }
        }
    }

    override fun emit(type: WsType, module: Module): File = file(type.identifier.value) {}
    override fun emit(enum: Enum, module: Module): File = file(enum.identifier.value) {}
    override fun emit(refined: Refined): File = file(refined.identifier.value) {}
    override fun emit(union: Union): File = file(union.identifier.value) {}
    override fun emit(channel: Channel): File = file(channel.identifier.value) {}

    private fun collectApplyParams(endpoint: Endpoint): List<Pair<String, Type>> = buildList {
        endpoint.path.filterIsInstance<Endpoint.Segment.Param>().forEach {
            add(it.identifier.value to it.reference.toIrType())
        }
        endpoint.queries.forEach { add(it.identifier.value to it.reference.toIrType()) }
        endpoint.headers.forEach { add(it.identifier.value to it.reference.toIrType()) }
        endpoint.requests.forEach { req ->
            val content = req.content ?: return@forEach
            add("body" to content.reference.toIrType())
        }
    }
}
